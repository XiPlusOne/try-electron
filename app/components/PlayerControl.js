import React from 'react';
import styles from './PlayerControl.less';

type Props = {
  playing: boolean,
  onToggleCallback: () => void
};

export default class PlayerControl extends React.PureComponent {
  props: Props;

  constructor(props) {
    super(props);

    this.onToggle = this.onToggle.bind(this);
  }

  onToggle() {
    const { onToggleCallback } = this.props;

    if (typeof onToggleCallback === 'function') onToggleCallback();
  }

  render() {
    const { playing } = this.props;

    return (
      <div className={styles.player}>
        <b className={`${styles.switch} ${styles.reverse}`} />
        <button
          type="button"
          onClick={this.onToggle}
          className={`${styles.toggle} ${
            playing ? styles.playing : styles.paused
          }`}
        />
        <b className={styles.switch} />
      </div>
    );
  }
}
