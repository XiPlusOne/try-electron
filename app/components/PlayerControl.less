.equilateral-triangle(@height) {
  @topsize : sqrt(pow(@height / 3 * 2, 2) - pow(@height / 3, 2));

  width: 0;
  height: 0;
  display: block;
  border-left: @height solid #fff;
  border-top: @topsize solid transparent;
  border-right: @height solid transparent;
  border-bottom: @topsize solid transparent;
}

.player {
  padding: 0 10px;
  margin: 0;
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 100%;

  .switch {
    @height: 15px;
    @borderTopSize: sqrt(pow(@height / 3 * 2, 2) - pow(@height / 3, 2));

    position: relative;
    .equilateral-triangle(@height);

    &::after {
      content: '';
      display: block;
      position: absolute;
      top: -@borderTopSize;
      left: 0;
      border-left: 1px solid #fff;
      height: 2 * @borderTopSize;
    }

    &.reverse {
      transform: rotate(180deg);
    }
  }

  .toggle {
    @radius: 15px;
    @range: 4px;
    @borderTopSize: sqrt(pow(@range * 2, 2) - pow(@range, 2));

    padding: 0;
    box-sizing: content-box;
    border: 1px solid #fff;
    width: 2 * @radius;
    height: 2 * @radius;
    border-radius: 50%;
    background: transparent;
    position: relative;
    outline: none;
    cursor: pointer;

    &:active {
      animation: flash 0.05s ease-out;
    }

    &::before {
      content: '';
      display: block;
      position: absolute;
    }

    &.paused::before {
      top: @radius - @borderTopSize;
      left: @radius - @range;
      .equilateral-triangle(@radius - @range);
    }

    &.playing::before {
      border-left: 1.5px solid #fff;
      border-right: 1.5px solid #fff;
      width: 5px;
      height: 10px;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
    }
  }
}

@keyframes flash {
  from {
    transform: scale(1.1);
  }

  to {
    transform: rotate(1);
  }
}
