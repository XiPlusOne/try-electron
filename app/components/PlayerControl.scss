.player {
  padding: 0 10px;
  margin: 0;
  list-style-type: none;
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 100%;

  .switch {
    $triangleHeight: 15px;
    // $borderTopSize² = ( $triangleHeight * 2 / 3 )² - ( $triangleHeight / 3 )²
    $borderTopSize: 8.66px;

    position: relative;
    width: 0px;
    height: 0px;
    border-left: $triangleHeight solid #fff;
    border-top: $borderTopSize solid transparent;
    border-right: $triangleHeight solid transparent;
    border-bottom: $borderTopSize solid transparent;

    &::after {
      content: '';
      display: block;
      position: absolute;
      top: -$borderTopSize;
      left: 0;
      border-left: 1px solid #fff;
      height: 2 * $borderTopSize;
    }

    &.reverse {
      transform: rotate(180deg);
    }
  }

  .toggle {
    $radius: 15px;
    $range: 4px;
    // $borderTopSize² = ( 2 * $range )² - $range²
    // 因为Sass不支持开平方运算符，所以在这里先算好
    $borderTopSize: 6.93px;

    padding: 0;
    box-sizing: content-box;
    border: 1px solid #fff;
    width: 2 * $radius;
    height: 2 * $radius;
    border-radius: 50%;
    background: transparent;
    position: relative;
    outline: none;
    cursor: pointer;

    &:active {
      animation: flash 0.05s ease-out;
    }

    &::before {
      content: '';
      display: block;
      position: absolute;
    }

    &.paused::before {
      top: $radius - $borderTopSize;
      left: $radius - $range;
      width: 0px;
      height: 0px;
      border-left: $range * 3 solid #fff;
      border-top: $borderTopSize solid transparent;
      border-right: $range * 3 solid transparent;
      border-bottom: $borderTopSize solid transparent;
    }

    &.playing::before {
      border-left: 1.5px solid #fff;
      border-right: 1.5px solid #fff;
      width: 5px;
      height: 10px;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
    }
  }
}

@keyframes flash {
  from {
    transform: scale(1.1);
  }

  to {
    transform: rotate(1);
  }
}
