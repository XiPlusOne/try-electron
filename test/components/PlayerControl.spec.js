import { spy } from 'sinon';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import PlayerControl from '../../app/components/PlayerControl';

Enzyme.configure({ adapter: new Adapter() });

describe('PlayerControl component', () => {
  it('should match exact snapshot', () => {
    const onToggleCallback = spy();
    const playerControl = (
      <PlayerControl playing={false} onToggleCallback={onToggleCallback} />
    );
    const tree = renderer.create(playerControl).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should toggle button should call onToggleCallback', () => {
    const onToggleCallback = spy();
    const component = shallow(
      <PlayerControl playing={false} onToggleCallback={onToggleCallback} />
    );

    const toggleButton = component.find('button').at(0);
    expect(toggleButton.hasClass('paused')).toBe(true);
    toggleButton.simulate('click');
    expect(onToggleCallback.called).toBe(true);
  });

  it('should toggle button should have a class named paused', () => {
    const component = shallow(<PlayerControl playing={false} />);
    const toggleButton = component.find('button').at(0);

    expect(toggleButton.hasClass('paused')).toBe(true);
  });

  it('should toggle button should have a class named playing', () => {
    const component = shallow(<PlayerControl playing />);
    const toggleButton = component.find('button').at(0);

    expect(toggleButton.hasClass('playing')).toBe(true);
  });
});
