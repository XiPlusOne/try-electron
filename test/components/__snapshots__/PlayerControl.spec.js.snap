// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerControl component should match exact snapshot 1`] = `
<div
  className="player"
>
  <b
    className="switch reverse"
  />
  <button
    className="toggle paused"
    onClick={[Function]}
    type="button"
  />
  <b
    className="switch"
  />
</div>
`;
